sfepy/solvers/ts_controllers.py

```diff
@@ -0,0 +1,229 @@
+"""
+Time step controllers.
+"""
+import numpy as nm
+
+from sfepy.base.base import Struct
+from sfepy.solvers.solvers import TimeStepController
+
+class FixedTSC(TimeStepController):
+    """
+    Fixed (do-nothing) time step controller.
+    """
+    name = 'tsc.fixed'
+
+class TimesSequenceTSC(TimeStepController):
+    """
+    Given times sequence time step controller.
+    """
+    name = 'tsc.time_sequence'
+
+    _parameters = [
+        ('times', 'iterable', range(1, 6), True,
+         'A sequence of times to generate.'),
+    ]
+
+    def __init__(self, conf, **kwargs):
+        TimeStepController.__init__(self, conf=conf, **kwargs)
+
+        self.iter_times = iter(self.conf.times)
+
+    def get_initial_dt(self, ts, vec, **kwargs):
+        # This cannot be called repeatedly!
+        self.t0 = ts.t0
+        return next(self.iter_times) - self.t0
+
+    def __call__(self, ts, vec0, vec1, **kwargs):
+        self.t0 = ts.time
+        try:
+            t1 = next(self.iter_times)
+
+        except StopIteration:
+            t1 = ts.t1
+
+        new_dt = t1 - self.t0
+        if new_dt == 0.0:
+            new_dt = 1.0
+
+        status = Struct(u_err=None, v_err=None, emax=None, result='accept')
+
+        return new_dt, status
+
+def eval_scaled_norm(terr, eps_a, eps_r):
+    return nm.sqrt(
+        1 / len(terr) *
+        nm.sum((terr / (eps_r * nm.abs(terr) + eps_a))**2)
+    )
+
+class ElastodynamicsBasicTSC(TimeStepController):
```

Comment:
Yes. this controller uses explicitly errors in u and v and is for the second order ODEs. It could be generalized for coupled multiphysical problems with both first and second ODEs - maybe later, as need arises.